kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/one.kt

```diff
@@ -0,0 +1,52 @@
+package io.kotest.assertions
+
+import io.kotest.common.ExperimentalKotest
+
+/**
+ * Executes the given lambda and expects exactly one assertion to succeed.
+ * If zero, two or more assertions succeed then this will cause this block to fail.
+ *
+ * ```
+ *   one {
+ *      "foo" shouldBe "bar"
+ *      "foo" shouldBe "foo
+ *      "foo" shouldBe "baz"
+ *   }
+ * ```
+ */
+@ExperimentalKotest
+suspend fun <T> one(assertions: suspend () -> T): T? {
+   val (result, failures, assertionsCount) = errorAndAssertionsScope { assertions() }
+   assertionCounter.inc(assertionsCount)
+
+   if (assertionsCount < 2) {
+      errorCollector.collectOrThrow(failures + failure("One cannot ensure a mutual exclusion with less than two assertions"))
+   }
+
+   if (assertionsCount == failures.size + 1) {
+      return result.getOrNull()
+   }
+
+   errorCollector.pushErrors(failures)
+
+   val f = when {
+      assertionsCount == failures.size -> failure("One expected a single assertion to succeed, but none succeeded.")
+      assertionsCount > failures.size + 1 -> failure("One expected a single assertion to succeed, but more than one succeeded.")
+      else -> failure("One expected a single assertion to succeed, but there were more failures than assertions.")
+   }
+
+   errorCollector.pushErrorAndMaybeThrow(f)
+   return null
```

Comment:
Would we ever have a situation where one wouldn't either throw or return the value ?

Reply:
I think if the assertion mode is soft, but also I think we need it to make the compiler happy.